corehq/tests/locks.py

```diff
@@ -0,0 +1,80 @@
+import logging
+from contextlib import contextmanager
+from threading import RLock
+from unittest.mock import patch
+
+import attr
+
+import dimagi.utils.couch as module
+
+log = logging.getLogger(__name__)
+
+
+def reentrant_redis_locks(test=None):
+    """Decorator/context manager to enable reentrant redis locks
+
+    This is useful for tests that do things like acquire a lock and then
+    fire off a celery task (which will usually be executed synchronously
+    in tests) before the lock is released.
```

Comment:
Scanning the code it looks like this basically loosens the behavior to say it's okay for a lock to be acquired multiple times at once, as long as _by the end_ when the context manager exists, no one's still holding the lock.

That also jives with the wikipedia definition of reentrant locks https://en.wikipedia.org/wiki/Reentrant_mutex (which yes I did have to look up) in case that helps anyone else as well.

Daniel, anything to add to how I described it, or is that pretty much right?

Reply:
Yes,  what you said is correct. I will update this docstring a bit since it's not entirely clear as is (blame it on writing docstrings hastily at 10:00pm). I'll improve it in a follow-up PR.

Reply:
https://github.com/dimagi/commcare-hq/pull/27402/commits/3f2cc9c05c43544bc66a5e2d358bc802c2e3fe02